runtime/near-vm-logic/src/logic.rs

```diff
@@ -943,14 +945,117 @@ impl<'a> VMLogic<'a> {
     pub fn keccak512(&mut self, value_len: u64, value_ptr: u64, register_id: u64) -> Result<()> {
         self.gas_counter.pay_base(keccak512_base)?;
         let value = self.get_vec_from_memory_or_register(value_ptr, value_len)?;
-        self.gas_counter.pay_per_byte(keccak512_byte, value.len() as u64)?;
+        self.gas_counter.pay_per(keccak512_byte, value.len() as u64)?;
 
         use sha3::Digest;
 
         let value_hash = sha3::Keccak512::digest(&value);
         self.internal_write_register(register_id, value_hash.as_slice().to_vec())
     }
 
+    /// Hashes the given value using RIPEMD-160 and returns it into `register_id`.
+    ///
+    /// # Errors
+    ///
+    /// If `value_len + value_ptr` points outside the memory or the registers use more memory than
+    /// the limit with `MemoryAccessViolation`.
+    ///
+    /// # Cost
+    ///
+    ///  Where `message_blocks` is `(value_len + 9).div_ceil(64)`.
+    ///
+    /// `base + write_register_base + write_register_byte * num_bytes + ripemd160_base + ripemd160_block * message_blocks`
+    #[cfg(feature = "protocol_feature_math_extension")]
+    pub fn ripemd160(&mut self, value_len: u64, value_ptr: u64, register_id: u64) -> Result<()> {
+        self.gas_counter.pay_base(ripemd160_base)?;
+        let value = self.get_vec_from_memory_or_register(value_ptr, value_len)?;
+
+        let message_blocks = value
+            .len()
+            .checked_add(8)
+            .ok_or(VMLogicError::HostError(HostError::IntegerOverflow))?
+            / 64
+            + 1;
+
+        self.gas_counter.pay_per(ripemd160_block, message_blocks as u64)?;
+
+        use ripemd160::Digest;
+
+        let value_hash = ripemd160::Ripemd160::digest(&value);
+        self.internal_write_register(register_id, value_hash.as_slice().to_vec())
+    }
+
+    /// Recovers an ECDSA signer address and returns it into `register_id`.
+    ///
+    /// Takes in an additional flag to check for malleability of the signature
+    /// which is generally only ideal for transactions.
+    ///
+    /// Returns a bool indicating success or failure as a `u64`.
+    ///
+    /// # Malleability Flags
+    ///
+    /// 0 - No malleability check.
+    /// 1 - Check malleability.
+    /// 2 - Rejecting upper range.
+    ///
+    /// # Errors
+    ///
+    /// * If `hash_ptr`, `r_ptr`, or `s_ptr` point outside the memory or the registers use more
+    ///   memory than the limit, then returns `MemoryAccessViolation`.
+    ///
+    /// # Cost
+    ///
+    /// `base + write_register_base + write_register_byte * 64 + ecrecover_base`
+    #[cfg(feature = "protocol_feature_math_extension")]
+    pub fn ecrecover(
+        &mut self,
+        hash_ptr: u64,
+        sig_ptr: u64,
+        v: u64,
+        malleability_flag: u64,
```

Comment:
Please combine them into a single `flags` parameter: low 2 bits are `v`, next bit is malleability check flag, remaining bits must be zero.

Reply:
Besides the other feedback, this is the only one that I am a bit confused about. I'm struggling to understand why you want to combine a flag and a parameter? Out of curiosity, why?

Reply:
Well, it sounds kinda wasteful to have 2 64-bit parameters for 3 bits of data. Also, if the hash and the signature are passed using register API (so each of them is represented by two parameters because you cannot omit the length because it is needed to distinguish a memory argument from a register argument), there would be 5 other parameters, and I don't like having too many parameters. This is not required, though.

Reply:
I do absolutely agree that it is wasteful but not quite uncommon across the other host functions. I also do want to button down the API and not have to change it in the future. 5 is usually my limit, I hate going over 5. I do see others using `u32`, Max I think suggested `u64` instead. I know we can't do `u8`.

Reply:
@bowenwang1996 Do you have thoughts on this?

Reply:
Bowen informed me that he is good with this.

Reply:
I suggest we keep arguments separately, it is better to have cleaner API than do micro-optimizations.